File: python/2021/03/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def read_input():
    return u.read_lines(file)

def part1():
    A = read_input()

    size = len(A[0])
    c = [0 for i in range(size)]
    for a in A:
        for i in range(size):
            c[i] += int(a[i])
    g = [1 if s > len(A)//2 else 0 for s in c]
    e = [0 if s > len(A)//2 else 1 for s in c]
    
    gamma = 0
    for x in g:
        gamma = gamma * 2 + x

    epsilon = 0
    for x in e:
        epsilon = epsilon * 2 + x

    k = gamma*epsilon
    print("Part 1:",k)

def part2():
    A = read_input()
    
    size = len(A[0])
    c = [0 for i in range(size)]

    for i in range(size):
        for a in A:
            c[i] += int(a[i])
        A = [a for a in A if a[i] == ("1" if c[i] >= len(A)/2 else "0")]
        if len(A) == 1:
            break
    oxygen = A[0]

    A = read_input()
    c = [0 for i in range(size)]
    for i in range(size):
        for a in A:
            c[i] += int(a[i])
        A = [a for a in A if a[i] == ("0" if c[i] >= len(A)/2 else "1")]
        if len(A) == 1:
            break
    co2 = A[0]

    k = int(oxygen,2)*int(co2,2)
    print("Part 2:",k)

part1()
part2()

File: c/2021/03/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif


void part1() {
    ARR_STRING* in = readInput(FILE_NAME);
    int64_t k = 0, l = 0;
    for (int16_t i=0; i<strlen(in->strings[0]); i++) {
	int64_t ones = 0;
    	for (int16_t j=0; j<in->length; j++) {
	    if (in->strings[j][i] == '1') ones++;
	}
	k = k << 1 | (ones > in->length/2 ? 1 : 0);
	l = l << 1 | (ones <= in->length/2 ? 1 : 0);
    }
    printf("Part 1: %ld\n", k*l);
    freeArrString(in);
}

int p(ARR_STRING* in, int8_t* good, char c, char d) {
    for (int16_t i=0; i<strlen(in->strings[0]); i++) {
	int64_t ones = 0;
	int64_t size = 0;
    	for (int16_t j=0; j<in->length; j++) {
	    if (good[j] == 0) continue;
	    size++;
	    if (in->strings[j][i] == '1') ones++;
	}
	if (size == 1) break;
	char q = ones >= size/2.0f ? c : d;
    	for (int16_t j=0; j<in->length; j++) {
	    if (in->strings[j][i] != q) good[j] = 0;
	}
    }
    char *kk = malloc(100);
    for (int64_t n = 0; n < in->length; n++) {
	if (good[n] > 0) {
	    kk = strcpy(kk, in->strings[n]);
	    break;
	}
    }
    int64_t k = 0;
    for (int64_t i = 0; i < strlen(kk); i++) {
    	k = k << 1 | (kk[i] == '1' ? 1 : 0);
    }
    free(kk);
    return k;
}

void part2() {
    ARR_STRING* in = readInput(FILE_NAME);
    int64_t k = 0, l = 0;
    int8_t* good = malloc(sizeof(int8_t)*in->length);
    memset(good, 1, in->length);
    k = p(in, good, '1', '0');
    memset(good, 1, in->length); 
    l = p(in, good, '0', '1');
    printf("Part 2: %ld\n", k*l);
    free(good);
    freeArrString(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}
